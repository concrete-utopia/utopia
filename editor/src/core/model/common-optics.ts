import type { EditorStorePatched } from '../../components/editor/store/editor-state'
import {
  modifyUnderlyingForOpenFile,
  withUnderlyingTargetFromEditorState,
} from '../../components/editor/store/editor-state'
import type { JSXElementChild } from '../shared/element-template'
import type { Optic } from '../shared/optics/optics'
import { traversal } from '../shared/optics/optics'
import type { ElementPath } from '../shared/project-file-types'

export function forElementOptic(target: ElementPath): Optic<EditorStorePatched, JSXElementChild> {
  function from(store: EditorStorePatched): Array<JSXElementChild> {
    return withUnderlyingTargetFromEditorState(target, store.editor, [], (_, element) => {
      return [element]
    })
  }
  function update(
    store: EditorStorePatched,
    modify: (child: JSXElementChild) => JSXElementChild,
  ): EditorStorePatched {
    return {
      ...store,
      editor: modifyUnderlyingForOpenFile(target, store.editor, modify),
    }
  }
  return traversal(from, update)
}
