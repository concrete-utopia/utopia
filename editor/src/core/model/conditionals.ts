import { ElementPath } from '../shared/project-file-types'
import * as EP from '../shared/element-path'
import {
  ElementInstanceMetadata,
  ElementInstanceMetadataMap,
  isJSExpression,
  isJSXConditionalExpression,
  isNullJSXAttributeValue,
  JSXConditionalExpression,
  JSXElementChild,
} from '../shared/element-template'
import { ElementPathTree, ElementPathTrees } from '../shared/element-path-tree'
import { getUtopiaID } from '../shared/uid-utils'
import { Optic } from '../shared/optics/optics'
import { fromField, fromTypeGuard } from '../shared/optics/optic-creators'
import { findUtopiaCommentFlag, isUtopiaCommentFlagConditional } from '../shared/comment-flags'
import { isLeft, isRight } from '../shared/either'
import { MetadataUtils } from './element-metadata-utils'
import { forceNotNull } from '../shared/optional-utils'

export type ConditionalCase = 'true-case' | 'false-case'

// Get the path for the clause (true case or false case) of a conditional.
export function getConditionalClausePath(
  conditionalPath: ElementPath,
  conditionalClause: JSXElementChild,
): ElementPath {
  return EP.appendToPath(conditionalPath, getUtopiaID(conditionalClause))
}

// Ensure that the children of a conditional are the whenTrue clause followed
// by the whenFalse clause.
export function reorderConditionalChildPathTrees(
  conditional: JSXConditionalExpression,
  conditionalPath: ElementPath,
  children: Array<ElementPathTree>,
): Array<ElementPathTree> {
  let result: Array<ElementPathTree> = []

  // The whenTrue clause should be first.
  const trueCasePath = getConditionalClausePath(conditionalPath, conditional.whenTrue)
  const trueCasePathString = EP.toString(trueCasePath)
  const trueCasePathTree = children.find((child) => child.pathString === trueCasePathString)
  if (trueCasePathTree != null) {
    result.push(trueCasePathTree)
  }

  // The whenFalse clause should be second.
  const falseCasePath = getConditionalClausePath(conditionalPath, conditional.whenFalse)
  const falseCasePathString = EP.toString(falseCasePath)
  const falseCasePathTree = children.find((child) => child.pathString === falseCasePathString)
  if (falseCasePathTree != null) {
    result.push(falseCasePathTree)
  }

  return result
}

export const jsxConditionalExpressionOptic: Optic<JSXElementChild, JSXConditionalExpression> =
  fromTypeGuard(isJSXConditionalExpression)

export const conditionalWhenTrueOptic: Optic<JSXConditionalExpression, JSXElementChild> =
  fromField('whenTrue')

export const conditionalWhenFalseOptic: Optic<JSXConditionalExpression, JSXElementChild> =
  fromField('whenFalse')

export function getClauseOptic(
  clause: ConditionalCase,
): Optic<JSXConditionalExpression, JSXElementChild> {
  if (clause === 'true-case') {
    return conditionalWhenTrueOptic
  } else {
    return conditionalWhenFalseOptic
  }
}

export function getConditionalFlag(element: JSXConditionalExpression): boolean | null {
  const flag = findUtopiaCommentFlag(element.comments, 'conditional')
  if (!isUtopiaCommentFlagConditional(flag)) {
    return null
  }
  return flag.value
}

export function matchesOverriddenConditionalBranch(
  elementPath: ElementPath,
  parentPath: ElementPath,
  params: {
    clause: JSXElementChild
    wantOverride: boolean
    parentOverride: boolean
  },
): boolean {
  const { clause, wantOverride, parentOverride } = params
  return (
    wantOverride === parentOverride &&
    EP.pathsEqual(elementPath, getConditionalClausePath(parentPath, clause))
  )
}

export function maybeConditionalExpression(
  element: ElementInstanceMetadata | null,
): JSXConditionalExpression | null {
  if (
    element != null &&
    isRight(element.element) &&
    isJSXConditionalExpression(element.element.value)
  ) {
    return element.element.value
  }
  return null
}

export function findMaybeConditionalExpression(
  elementPath: ElementPath | null,
  jsxMetadata: ElementInstanceMetadataMap,
): JSXConditionalExpression | null {
  return maybeConditionalExpression(
    MetadataUtils.findElementByElementPath(jsxMetadata, elementPath),
  )
}

export function maybeBranchConditionalCase(
  conditionalPath: ElementPath,
  conditional: JSXConditionalExpression | null,
  branchPath: ElementPath,
): ConditionalCase | null {
  if (conditional == null) {
    return null
  }
  const truePath = EP.appendToPath(conditionalPath, conditional.whenTrue.uid)
  const falsePath = EP.appendToPath(conditionalPath, conditional.whenFalse.uid)
  if (EP.pathsEqual(truePath, branchPath)) {
    return 'true-case'
  } else if (EP.pathsEqual(falsePath, branchPath)) {
    return 'false-case'
  } else {
    return null
  }
}

export function maybeConditionalActiveBranch(
  elementPath: ElementPath | null,
  jsxMetadata: ElementInstanceMetadataMap,
): JSXElementChild | null {
  if (elementPath == null) {
    return null
  }
  const conditional = maybeConditionalExpression(
    MetadataUtils.findElementByElementPath(jsxMetadata, elementPath),
  )
  if (conditional == null) {
    return null
  }

  const activeCase = forceNotNull(
    'conditional should have an active case',
    getConditionalActiveCase(elementPath, conditional, jsxMetadata),
  )
  return getConditionalBranch(conditional, activeCase)
}

export function getConditionalCaseCorrespondingToBranchPath(
  branchPath: ElementPath,
  metadata: ElementInstanceMetadataMap,
): ConditionalCase | null {
  const conditionalElement = findMaybeConditionalExpression(EP.parentPath(branchPath), metadata)
  if (conditionalElement == null) {
    return null
  }

  return maybeBranchConditionalCase(EP.parentPath(branchPath), conditionalElement, branchPath)
}

export function getConditionalBranch(
  conditional: JSXConditionalExpression,
  clause: ConditionalCase,
): JSXElementChild {
  return clause === 'true-case' ? conditional.whenTrue : conditional.whenFalse
}

export function isConditionalWithEmptyActiveBranch(
  path: ElementPath,
  metadata: ElementInstanceMetadataMap,
): {
  element: JSXConditionalExpression
  clause: ConditionalCase | null
  isEmpty: boolean
} | null {
  const conditional = findMaybeConditionalExpression(path, metadata)
  if (conditional == null) {
    return null
  }
  const clause = getConditionalActiveCase(path, conditional, metadata)
  if (clause == null) {
    return {
      element: conditional,
      isEmpty: false,
      clause,
    }
  }
  const branch = clause === 'true-case' ? conditional.whenTrue : conditional.whenFalse
  return {
    element: conditional,
    isEmpty: isNullJSXAttributeValue(branch),
    clause,
  }
}

export function isNonEmptyConditionalBranch(
  elementPath: ElementPath,
  jsxMetadata: ElementInstanceMetadataMap,
): boolean {
  const parentPath = EP.parentPath(elementPath)
  const conditionalParent = findMaybeConditionalExpression(parentPath, jsxMetadata)
  if (conditionalParent == null) {
    return false
  }
  const clause = maybeBranchConditionalCase(parentPath, conditionalParent, elementPath)
  if (clause == null) {
    return false
  }
  const branch = getConditionalBranch(conditionalParent, clause)
  return !isNullJSXAttributeValue(branch)
}

export function isEmptyConditionalBranch(
  elementPath: ElementPath,
  jsxMetadata: ElementInstanceMetadataMap,
): boolean {
  return !isNonEmptyConditionalBranch(elementPath, jsxMetadata)
}

export function getConditionalClausePathFromMetadata(
  conditionalPath: ElementPath,
  metadata: ElementInstanceMetadataMap,
  clause: ConditionalCase,
): ElementPath | null {
  const conditionalElement = findMaybeConditionalExpression(conditionalPath, metadata)
  if (conditionalElement == null) {
    return null
  }

  return getConditionalClausePath(
    conditionalPath,
    clause === 'true-case' ? conditionalElement.whenTrue : conditionalElement.whenFalse,
  )
}

export function isOverriddenConditional(element: ElementInstanceMetadata | null): boolean {
  if (
    element == null ||
    isLeft(element.element) ||
    !isJSXConditionalExpression(element.element.value)
  ) {
    return false
  }
  return getConditionalFlag(element.element.value) != null
}

export function getConditionalActiveCase(
  path: ElementPath,
  conditional: JSXConditionalExpression,
  spyMetadata: ElementInstanceMetadataMap,
): ConditionalCase | null {
  const override = getConditionalFlag(conditional)
  if (override != null) {
    return override ? 'true-case' : 'false-case'
  }
  const spy = spyMetadata[EP.toString(path)] ?? null
  if (spy == null) {
    return 'true-case'
  }
  if (spy.conditionValue === 'not-a-conditional') {
    return null
  }
  return spy.conditionValue.active ? 'true-case' : 'false-case'
}

export function conditionalClauseAsBoolean(clause: ConditionalCase): boolean {
  return clause === 'true-case'
}

export function isActiveBranchOfConditional(
  clause: ConditionalCase,
  metadata: ElementInstanceMetadata | null,
): boolean {
  const conditionValue = metadata?.conditionValue
  if (conditionValue == null || conditionValue === 'not-a-conditional') {
    return false
  } else {
    const clauseValue = conditionalClauseAsBoolean(clause)
    return clauseValue === conditionValue.active
  }
}

export function isDefaultBranchOfConditional(
  clause: ConditionalCase,
  metadata: ElementInstanceMetadata | null,
): boolean {
  const conditionValue = metadata?.conditionValue
  if (conditionValue == null || conditionValue === 'not-a-conditional') {
    return false
  } else {
    const clauseValue = conditionalClauseAsBoolean(clause)
    return clauseValue === conditionValue.default
  }
}

export function isActiveOrDefaultBranchOfConditional(
  clause: ConditionalCase,
  metadata: ElementInstanceMetadata | null,
): boolean {
  return (
    isActiveBranchOfConditional(clause, metadata) || isDefaultBranchOfConditional(clause, metadata)
  )
}

export function findFirstNonConditionalAncestor(
  initial: ElementPath,
  metadata: ElementInstanceMetadataMap,
): ElementPath {
  const parent = EP.parentPath(initial)
  if (findMaybeConditionalExpression(parent, metadata) == null) {
    return parent
  }
  return findFirstNonConditionalAncestor(parent, metadata)
}

export function isTextEditableConditional(
  path: ElementPath,
  metadata: ElementInstanceMetadataMap,
  elementPathTree: ElementPathTrees,
): boolean {
  const element = MetadataUtils.findElementByElementPath(metadata, path)
  if (element == null) {
    // element doesn't exist
    return false
  }
  const conditional = maybeConditionalExpression(element)
  if (conditional == null) {
    // element is not a conditional
    return false
  }
  const nonConditionalAncestor = findFirstNonConditionalAncestor(path, metadata)
  const siblings = MetadataUtils.getChildrenOrdered(
    metadata,
    elementPathTree,
    nonConditionalAncestor,
  )
  if (siblings.length !== 1) {
    // we don't allow text editing of conditional branches when the conditional has siblings
    // (or if the topmost nested conditional has siblings)
    return false
  }

  const childrenInMetadata = MetadataUtils.getChildrenUnordered(metadata, path)
  if (childrenInMetadata.length > 0) {
    // we don't allow text editing of conditional branches when the conditional have children in metadata
    // that would mean there are elefants in the active branch, which should not be text editable
    return false
  }

  // Finally we check if the active conditional branch is a js expression. Maybe this is an overkill,
  // because the earlier checks already prove this is a leaf element.
  const activeConditionalBranch = maybeConditionalActiveBranch(path, metadata)
  return activeConditionalBranch != null && isJSExpression(activeConditionalBranch)
}
