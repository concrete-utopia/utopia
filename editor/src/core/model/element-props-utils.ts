import { type AllElementProps } from '../../components/editor/store/editor-state'
import { type ElementPath, type PropertyPath } from '../shared/project-file-types'
import * as PP from '../shared/property-path'
import * as EP from '../shared/element-path'
import * as ObjectPath from 'object-path'

export function getElementPropValue(
  allElementProps: AllElementProps,
  target: ElementPath,
  propPath: PropertyPath,
): unknown {
  const propsForElement = allElementProps[EP.toString(target)] ?? {}
  return ObjectPath.get(propsForElement, PP.getElements(propPath))
}

export function elementHasValueForProp(
  allElementProps: AllElementProps,
  target: ElementPath,
  propPath: PropertyPath,
): boolean {
  return getElementPropValue(allElementProps, target, propPath) != null
}
