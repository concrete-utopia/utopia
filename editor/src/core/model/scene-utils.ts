import Hash from 'object-hash'
import {
  SceneMetadata,
  StaticElementPath,
  PropertyPath,
  isTextFile,
  isParseSuccess,
} from '../shared/project-file-types'
import {
  UtopiaJSXComponent,
  utopiaJSXComponent,
  jsxElement,
  JSXElement,
  jsExpressionValue,
  isJSXElement,
  JSXElementChild,
  TopLevelElement,
  isUtopiaJSXComponent,
  JSXAttributes,
  defaultPropsParam,
  jsExpressionOtherJavaScript,
  ElementInstanceMetadataMap,
  jsxAttributesFromMap,
  ElementInstanceMetadata,
  walkElements,
  emptyComments,
} from '../shared/element-template'
import * as EP from '../shared/element-path'
import * as PP from '../shared/property-path'
import {
  eitherToMaybe,
  applicative5Either,
  flatMapEither,
  isLeft,
  applicative6Either,
  right,
} from '../shared/either'
import { memoize } from '../shared/memoize'
import fastDeepEqual from 'fast-deep-equal'
import {
  getModifiableJSXAttributeAtPath,
  jsxSimpleAttributeToValue,
} from '../shared/jsx-attributes'
import { stripNulls } from '../shared/array-utils'
import { UTOPIA_UID_KEY } from './utopia-constants'
import { getContentsTreeFileFromString, ProjectContentTreeRoot } from '../../components/assets'
import { getUtopiaJSXComponentsFromSuccess } from './project-file-utils'
import { generateConsistentUID, generateUID, getUtopiaID } from '../shared/uid-utils'
import { emptySet } from '../shared/set-utils'

export const PathForSceneComponent = PP.create('component')
export const PathForSceneDataUid = PP.create('data-uid')
export const PathForSceneDataLabel = PP.create('data-label')
export const PathForSceneFrame = PP.create('style')

export const BakedInStoryboardUID = 'utopia-storyboard-uid'
export const BakedInStoryboardVariableName = 'storyboard'

export const EmptyUtopiaCanvasComponent = convertScenesToUtopiaCanvasComponent([])

export const PathForSceneProps = PP.create('props')
export const PathForSceneStyle = PP.create('style')

export function createSceneUidFromIndex(sceneIndex: number): string {
  return `scene-${sceneIndex}`
}

export function mapScene(scene: SceneMetadata): JSXElement {
  const sceneProps = jsxAttributesFromMap({
    component: jsExpressionOtherJavaScript(
      scene.component ?? 'null',
      `return ${scene.component}`,
      [],
      null,
      {},
    ),
    props: jsExpressionValue(scene.props, emptyComments),
    style: jsExpressionValue(scene.frame, emptyComments),
    'data-uid': jsExpressionValue(scene.uid, emptyComments),
    'data-label': jsExpressionValue(scene.label, emptyComments),
  })
  return jsxElement('Scene', scene.uid, sceneProps, [])
}

export function unmapScene(element: JSXElementChild): SceneMetadata | null {
  if (!isJSXElement(element) || element.name.baseVariable !== 'Scene') {
    return null
  }
  return eitherToMaybe(
    applicative6Either(
      (component, props, style, label, uid) => {
        const scene: SceneMetadata = {
          uid: uid,
          component: component,
          props: props,
          frame: style,
          ...(label == null ? {} : { label: label }),
        }
        return scene
      },
      getSimpleAttributeAtPathCustom(element.props, PP.create('component')),
      getSimpleAttributeAtPathCustom(element.props, PP.create('props')),
      getSimpleAttributeAtPathCustom(element.props, PP.create('style')),
      getSimpleAttributeAtPathCustom(element.props, PP.create('layout')),
      getSimpleAttributeAtPathCustom(element.props, PP.create('data-label')),
      getSimpleAttributeAtPathCustom(element.props, PP.create('data-uid')),
    ),
  )
}

export function convertScenesToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata>,
): UtopiaJSXComponent
export function convertScenesToUtopiaCanvasComponent(scenes: null): null
export function convertScenesToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata> | null,
): UtopiaJSXComponent | null
export function convertScenesToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata> | null,
): UtopiaJSXComponent | null {
  if (scenes == null) {
    return null
  }
  return utopiaJSXComponent(
    BakedInStoryboardVariableName,
    false,
    'var',
    'block',
    null,
    [],
    jsxElement(
      'Storyboard',
      BakedInStoryboardUID,
      jsxAttributesFromMap({ 'data-uid': jsExpressionValue(BakedInStoryboardUID, emptyComments) }),
      scenes.map(mapScene),
    ),
    null,
    false,
    emptyComments,
  )
}

export function createSceneFromComponent(
  filePath: string,
  componentImportedAs: string,
  uid: string,
): JSXElement {
  const sceneProps = jsxAttributesFromMap({
    [UTOPIA_UID_KEY]: jsExpressionValue(uid, emptyComments),
    style: jsExpressionValue(
      {
        position: 'absolute',
        left: 0,
        top: 0,
        width: 375,
        height: 812,
      },
      emptyComments,
    ),
  })
  const hash = Hash({
    fileName: filePath,
    name: componentImportedAs,
    props: jsxAttributesFromMap({}),
  })
  const componentUID = generateConsistentUID(hash)
  return jsxElement('Scene', uid, sceneProps, [
    jsxElement(
      componentImportedAs,
      componentUID,
      jsxAttributesFromMap({
        [UTOPIA_UID_KEY]: jsExpressionValue(componentUID, emptyComments),
      }),
      [],
    ),
  ])
}

export function createStoryboardElement(scenes: Array<JSXElement>, uid: string): JSXElement {
  const storyboardProps = jsxAttributesFromMap({
    [UTOPIA_UID_KEY]: jsExpressionValue(uid, emptyComments),
  })
  return jsxElement('Storyboard', uid, storyboardProps, scenes)
}

export function convertUtopiaCanvasComponentToScenes(
  utopiaCanvasComponent: UtopiaJSXComponent | null,
): Array<SceneMetadata> | null {
  if (utopiaCanvasComponent == null) {
    return null
  }
  const rootElement = utopiaCanvasComponent.rootElement
  if (!isJSXElement(rootElement) || rootElement.name.baseVariable !== 'Storyboard') {
    throw new Error('the root element must be a Storyboard component')
  }
  const firstChildrenScenes: Array<SceneMetadata> = stripNulls(rootElement.children.map(unmapScene))
  return firstChildrenScenes
}

const convertScenesAndTopLevelElementsToUtopiaCanvasComponentMemoized = memoize(
  (scenes: Array<SceneMetadata>, topLevelElements: Array<TopLevelElement>) => [
    ...topLevelElements,
    convertScenesToUtopiaCanvasComponent(scenes),
  ],
  { equals: fastDeepEqual }, // delete me as soon as possible
)

export function convertScenesAndTopLevelElementsToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata>,
  topLevelElements: Array<UtopiaJSXComponent>,
): Array<UtopiaJSXComponent>
export function convertScenesAndTopLevelElementsToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata>,
  topLevelElements: Array<TopLevelElement>,
): Array<TopLevelElement>
export function convertScenesAndTopLevelElementsToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata>,
  topLevelElements: Array<TopLevelElement>,
): Array<TopLevelElement> {
  return convertScenesAndTopLevelElementsToUtopiaCanvasComponentMemoized(scenes, topLevelElements)
}

export function fishOutUtopiaCanvasFromTopLevelElements(
  topLevelElements: Array<TopLevelElement>,
): UtopiaJSXComponent | null {
  return (
    topLevelElements.find((e): e is UtopiaJSXComponent => {
      return isUtopiaJSXComponent(e) && e.name === BakedInStoryboardVariableName
    }) ?? null
  )
}

function getSimpleAttributeAtPathCustom(attributes: JSXAttributes, path: PropertyPath) {
  const getAttrResult = getModifiableJSXAttributeAtPath(attributes, path)
  return flatMapEither((attr) => {
    const simpleValue = jsxSimpleAttributeToValue(attr)
    if (isLeft(simpleValue) && attr.type === 'ATTRIBUTE_OTHER_JAVASCRIPT') {
      return right(attr.javascript)
    } else {
      return simpleValue
    }
  }, getAttrResult)
}

export function sceneMetadata(
  uid: string,
  component: string | null,
  props: { [key: string]: any },
  frame: {
    left: number
    top: number
    width: number
    height: number
  },
  label?: string,
): SceneMetadata {
  let scene: SceneMetadata = {
    uid: uid,
    component: component,
    frame: frame,
    props: props,
  }

  if (label != null) {
    // This is annoying, but if we don't do this then we'll explicitly print
    // `undefined` in the label field, meaning the JSON won't parse
    scene.label = label
  }

  return scene
}

export function getStoryboardUID(openComponents: UtopiaJSXComponent[]): string | null {
  const possiblyStoryboard = openComponents.find(
    (component) => component.name === BakedInStoryboardVariableName,
  )
  if (possiblyStoryboard != null) {
    return getUtopiaID(possiblyStoryboard.rootElement)
  }
  return null
}

export function getStoryboardElementPath(
  projectContents: ProjectContentTreeRoot,
  openFile: string | null | undefined,
): StaticElementPath | null {
  if (openFile != null) {
    const file = getContentsTreeFileFromString(projectContents, openFile)
    if (file != null && isTextFile(file) && isParseSuccess(file.fileContents.parsed)) {
      const possiblyStoryboard = getUtopiaJSXComponentsFromSuccess(file.fileContents.parsed).find(
        (component) => component.name === BakedInStoryboardVariableName,
      )
      if (possiblyStoryboard != null) {
        const uid = getUtopiaID(possiblyStoryboard.rootElement)
        return EP.elementPath([EP.staticElementPath([uid])])
      }
    }
  }
  return null
}
