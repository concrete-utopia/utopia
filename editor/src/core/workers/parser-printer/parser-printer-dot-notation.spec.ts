import { applyPrettier } from './prettier-utils'
import {
  clearParseResultUniqueIDs,
  testParseCode,
  testParseThenPrint,
} from './parser-printer.test-utils'
import { BakedInStoryboardUID } from '../../model/scene-utils'
import { foldParsedTextFile } from '../../shared/project-file-types'

describe('JSX parser', () => {
  it('should parse out dot-notation element names', () => {
    const code = applyPrettier(
      `
import * as React from "react";
import * as Utopia from "utopia-api";
export var App = props => {
  return (
    <Utopia.View data-uid="aaa">
      {<div data-uid="bbb" />}
    </Utopia.View>
  )
}
export var storyboard = (props) => {
  return (
    <Storyboard data-uid='${BakedInStoryboardUID}'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        component={App}
        layout={{ layoutSystem: 'pinSystem' }}
        props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
        data-uid='scene-aaa'
      />
    </Storyboard>
  )
}
`,
      false,
    ).formatted
    const parseResult = clearParseResultUniqueIDs(testParseCode(code))
    foldParsedTextFile(
      (failure) => {
        fail(failure)
      },
      (success) => {
        expect(success.imports).toMatchSnapshot()
        expect(success.topLevelElements).toMatchSnapshot()
      },
      (unparsed) => {
        fail(unparsed)
      },
      parseResult,
    )
  })
  it('dot-notated elements cycle fully back and forth', () => {
    const originalCode = `
import * as React from "react";
import * as Utopia from "utopia-api";
import { Scene, Storyboard, View } from "utopia-api";
export var App = props => {
  return (
    <Utopia.View data-uid="aaa">
      <View data-uid='bbb' />
    </Utopia.View>
  )
}
export var storyboard = (props) => {
  return (
    <Storyboard data-uid='${BakedInStoryboardUID}'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        component={App}
        layout={{ layoutSystem: 'pinSystem' }}
        props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
        data-uid='scene-aaa'
      />
    </Storyboard>
  )
}`
    const printedCode = `import * as React from 'react'
import { Scene, Storyboard, View } from 'utopia-api'
import * as Utopia from 'utopia-api'
export var App = (props) => {
  return (
    <Utopia.View data-uid='aaa'>
      <View data-uid='bbb' />
    </Utopia.View>
  )
}
export var storyboard = (props) => {
  return (
    <Storyboard data-uid='${BakedInStoryboardUID}'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        component={App}
        layout={{ layoutSystem: 'pinSystem' }}
        props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
        data-uid='scene-aaa'
      />
    </Storyboard>
  )
}
`
    testParseThenPrint(originalCode, printedCode)
  })
})
