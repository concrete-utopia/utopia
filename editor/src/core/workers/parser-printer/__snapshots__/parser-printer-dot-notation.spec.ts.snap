// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX parser should parse out dot-notation element names 1`] = `
Object {
  "react": Object {
    "comments": Object {
      "leadingComments": Array [],
      "trailingComments": Array [],
    },
    "importedAs": "React",
    "importedFromWithin": Array [],
    "importedWithName": null,
  },
  "utopia-api": Object {
    "comments": Object {
      "leadingComments": Array [],
      "trailingComments": Array [],
    },
    "importedAs": "Utopia",
    "importedFromWithin": Array [],
    "importedWithName": null,
  },
}
`;

exports[`JSX parser should parse out dot-notation element names 2`] = `
Array [
  Object {
    "arbitraryJSBlock": null,
    "blockOrExpression": "block",
    "comments": Object {
      "leadingComments": Array [],
      "trailingComments": Array [],
    },
    "declarationSyntax": "var",
    "isFunction": true,
    "name": "App",
    "param": Object {
      "boundParam": Object {
        "defaultExpression": null,
        "paramName": "props",
        "type": "REGULAR_PARAM",
      },
      "dotDotDotToken": false,
      "type": "PARAM",
    },
    "propsUsed": Array [],
    "returnStatementComments": Object {
      "leadingComments": Array [],
      "trailingComments": Array [],
    },
    "rootElement": Object {
      "children": Array [
        Object {
          "definedElsewhere": Array [
            "React",
            "utopiaCanvasJSXLookup",
          ],
          "elementsWithin": Object {
            "bbb": Object {
              "children": Array [],
              "name": Object {
                "baseVariable": "div",
                "propertyPath": Object {
                  "propertyElements": Array [],
                },
              },
              "props": Object {
                "data-uid": Object {
                  "type": "ATTRIBUTE_VALUE",
                  "value": "bbb",
                },
              },
              "type": "JSX_ELEMENT",
            },
          },
          "javascript": "<div data-uid='bbb' />;",
          "originalJavascript": "<div data-uid='bbb' />",
          "sourceMap": Object {
            "file": "code.tsx",
            "mappings": "OAAA",
            "names": Array [],
            "sources": Array [
              "code.tsx",
            ],
            "sourcesContent": Array [
              "import * as React from 'react'
import * as Utopia from 'utopia-api'
export var App = (props) => {
  return <Utopia.View data-uid='aaa'>{<div data-uid='bbb' />}</Utopia.View>
}
export var storyboard = (props) => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        component={App}
        layout={{ layoutSystem: 'pinSystem' }}
        props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
        data-uid='scene-aaa'
      />
    </Storyboard>
  )
}
",
            ],
            "version": 3,
          },
          "transpiledJavascript": "return utopiaCanvasJSXLookup(\\"bbb\\", {});",
          "type": "JSX_ARBITRARY_BLOCK",
          "uniqueID": "",
        },
      ],
      "name": Object {
        "baseVariable": "Utopia",
        "propertyPath": Object {
          "propertyElements": Array [
            "View",
          ],
        },
      },
      "props": Object {
        "data-uid": Object {
          "type": "ATTRIBUTE_VALUE",
          "value": "aaa",
        },
      },
      "type": "JSX_ELEMENT",
    },
    "type": "UTOPIA_JSX_COMPONENT",
    "usedInReactDOMRender": false,
  },
  Object {
    "comments": Object {
      "leadingComments": Array [],
      "trailingComments": Array [],
    },
    "definedElsewhere": Array [
      "React",
      "Storyboard",
      "Scene",
      "App",
    ],
    "definedWithin": Array [
      "storyboard",
    ],
    "javascript": "export var storyboard = props => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
      style={{ left: 0, top: 0, width: 400, height: 400 }}
      component={App}
      layout={{ layoutSystem: 'pinSystem' }}
      props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
      data-uid='scene-aaa' />

    </Storyboard>);

};",
    "sourceMap": Object {
      "file": "code.tsx",
      "mappings": "AAKQA,IAAIC,UAAUC,GAAGC,SAAbF,UAAaE,CAACC,KAADD,EAAWE;AACjCC,SACEC,oBAACC,UAADD;AAAYE,gBAASC;AAArBH,KACEA,oBAACI,KAADJ;AACEK,IAAAA,KAAKC,EAAER;AAAES,MAAAA,IAAIC,EAAEC,CAARX;AAAWY,MAAAA,GAAGF,EAAEC,CAAhBX;AAAmBa,MAAAA,KAAKH,EAAEI,GAA1Bd;AAA+Be,MAAAA,MAAML,EAAEI;AAAvCd,KADTE;AAEEc,IAAAA,SAASR,EAAES,GAFbf;AAGEgB,IAAAA,MAAMV,EAAER;AAAEmB,MAAAA,YAAYT,EAAEL;AAAhBL,KAHVE;AAIEH,IAAAA,KAAKS,EAAER;AAAEkB,MAAAA,MAAMR,EAAEV;AAAEoB,QAAAA,MAAMV,EAAEC,CAAVX;AAAaS,QAAAA,IAAIC,EAAEC,CAAnBX;AAAsBqB,QAAAA,KAAKX,EAAEC,CAA7BX;AAAgCY,QAAAA,GAAGF,EAAEC;AAArCX;AAAVA,KAJTE;AAKEE,gBAASC;AALXH,IADFA,CADFD;AAWFqB,CAZO3B",
      "names": Array [
        "var",
        "storyboard",
        " ",
        "(",
        "props",
        "{",
        "return",
        "<",
        "Storyboard",
        "data",
        "'",
        "Scene",
        "style",
        "=",
        "left",
        ":",
        "0",
        "top",
        "width",
        "400",
        "height",
        "component",
        "App",
        "layout",
        "layoutSystem",
        "bottom",
        "right",
        "}",
      ],
      "sources": Array [
        "code.tsx",
      ],
      "sourcesContent": Array [
        "import * as React from 'react'
import * as Utopia from 'utopia-api'
export var App = (props) => {
  return <Utopia.View data-uid='aaa'>{<div data-uid='bbb' />}</Utopia.View>
}
export var storyboard = (props) => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        component={App}
        layout={{ layoutSystem: 'pinSystem' }}
        props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
        data-uid='scene-aaa'
      />
    </Storyboard>
  )
}
",
      ],
      "version": 3,
    },
    "transpiledJavascript": "var storyboard = function storyboard(props) {
  return React.createElement(Storyboard, {
    \\"data-uid\\": \\"utopia-storyboard-uid\\"
  }, React.createElement(Scene, {
    style: {
      left: 0,
      top: 0,
      width: 400,
      height: 400
    },
    component: App,
    layout: {
      layoutSystem: 'pinSystem'
    },
    props: {
      layout: {
        bottom: 0,
        left: 0,
        right: 0,
        top: 0
      }
    },
    \\"data-uid\\": \\"scene-aaa\\"
  }));
};
return { storyboard: storyboard };",
    "type": "ARBITRARY_JS_BLOCK",
    "uniqueID": "",
  },
]
`;
